crates/sp-domains-fraud-proof/src/verification.rs

```diff
@@ -458,3 +460,97 @@ where
         _ => Err(VerificationError::InvalidProof),
     }
 }
+
+/// Represents error for invalid bundle equivocation proof.
+#[derive(Debug)]
+#[cfg_attr(feature = "thiserror", derive(thiserror::Error))]
+pub enum InvalidBundleEquivocationError {
+    /// Bundle signature is invalid.
+    #[cfg_attr(feature = "thiserror", error("Invalid bundle signature."))]
+    BadBundleSignature,
+    /// Bundle slot mismatch.
+    #[cfg_attr(feature = "thiserror", error("Bundle slot mismatch."))]
+    BundleSlotMismatch,
+    /// Same bundle hash.
+    #[cfg_attr(feature = "thiserror", error("Same bundle hash."))]
+    SameBundleHash,
+    /// Invalid Proof of election.
+    #[cfg_attr(feature = "thiserror", error("Invalid Proof of Election: {0:?}"))]
+    InvalidProofOfElection(ProofOfElectionError),
+    /// Failed to get domain total stake.
+    #[cfg_attr(feature = "thiserror", error("Failed to get domain total stake."))]
+    FailedToGetDomainTotalStake,
+    /// Failed to get operator stake.
+    #[cfg_attr(feature = "thiserror", error("Failed to get operator stake"))]
+    FailedToGetOperatorStake,
+}
+
+/// Verifies Bundle equivocation fraud proof.
+#[allow(dead_code)]
+pub fn verify_bundle_equivocation_fraud_proof<CBlock, DomainHeader, Balance>(
+    operator_signing_key: &OperatorPublicKey,
+    header_1: SealedBundleHeader<NumberFor<CBlock>, CBlock::Hash, DomainHeader, Balance>,
+    header_2: SealedBundleHeader<NumberFor<CBlock>, CBlock::Hash, DomainHeader, Balance>,
+) -> Result<(), InvalidBundleEquivocationError>
+where
+    CBlock: BlockT,
+    DomainHeader: HeaderT,
+    Balance: Encode,
+{
+    if !operator_signing_key.verify(&header_1.pre_hash(), &header_1.signature) {
+        return Err(InvalidBundleEquivocationError::BadBundleSignature);
+    }
+
+    if !operator_signing_key.verify(&header_2.pre_hash(), &header_2.signature) {
+        return Err(InvalidBundleEquivocationError::BadBundleSignature);
+    }
+
+    let consensus_block_hash = header_1.header.proof_of_election.consensus_block_hash;
+    let domain_id = header_1.header.proof_of_election.domain_id;
+    let operator_id = header_1.header.proof_of_election.operator_id;
```

Comment:
Why?

Protocol does not support an operator for two domains at a single time. So they will not be able to create a valid PoE in this case. So Bundle validation will never go through.

What am I missing here?

Reply:
Sorry for not elaborating on it better!

Yes, we don't support an operator for two domains at a single time, but the user can use the same `signing_key` for 2 operators on different domain, if both operators win the same slot on different domains and produce 
2 bundles, then malicious operator can construct an invalid bundle equivocation proof that can pass the verification here.

Adding this check essentially ensure the bundle is produced by the same operator on the same slot.

Reply:
>  Yes, we don't support an operator for two domains at a single time, but the user can use the same signing_key for 2 operators on different domain, if both operators win the same slot on different domains and produce
2 bundles, then malicious operator can construct an invalid bundle equivocation proof that can pass the verification here.

Even if malicious operator constructs invalid bundle proof that is targeting the two different headers belonging to two different domains the POE check is assuming same domain which will fail if not the same domain.

Reply:
But I do like the idea of short circuiting earlier. Will add this check anyway

Reply:
> the POE check is assuming same domain which will fail if not the same domain

Not really, from my understanding, once the `verify_vrf_signature` is passed, the `proof_of_election.vrf_signature.output` is served as a randomness, and as long as the threshold calculated from `(operator_stake, total_domain_stake, bundle_slot_probability)` is large enough, it can pass the verification with high probability regardless what the randomness is.

Reply:
Hmm I see